src/mol-math/linear-algebra/tensor.ts

```diff
@@ -313,4 +313,198 @@ export namespace Tensor {
+                                }
+                            }
+                            if (vx < xn - 1) {
+                                const noffset = o(vx + 1, vy, vz);
+                                if (data[noffset] < threshold && !visited[noffset]) {
+                                    stack.push(noffset);
+                                }
+                            }
+                            if (vy > 0) {
+                                const noffset = o(vx, vy - 1, vz);
+                                if (data[noffset] < threshold && !visited[noffset]) {
+                                    stack.push(noffset);
+                                }
+                            }
+                            if (vy < yn - 1) {
+                                const noffset = o(vx, vy + 1, vz);
+                                if (data[noffset] < threshold && !visited[noffset]) {
+                                    stack.push(noffset);
+                                }
+                            }
+                            if (vz > 0) {
+                                const noffset = o(vx, vy, vz - 1);
+                                if (data[noffset] < threshold && !visited[noffset]) {
+                                    stack.push(noffset);
+                                }
+                            }
+                            if (vz < zn - 1) {
+                                const noffset = o(vx, vy, vz + 1);
+                                if (data[noffset] < threshold && !visited[noffset]) {
+                                    stack.push(noffset);
+                                }
+                            }
+                        }
+                    }
+                }
+            }
+        }
+
+        return visited;
+    }
+
+    /**
+     * Floodfill from boundary voxels below the given threshold.
+     * Only for rank-1, rank-2, and rank-3 tensors.
+     */
+    export function floodfill(tensor: Tensor, threshold: number): Uint8Array {
+        switch (tensor.space.dimensions.length) {
+            case 1: return floodfill1(tensor, threshold);
+            case 2: return floodfill2(tensor, threshold);
+            case 3: return floodfill3(tensor, threshold);
+            default: throw new Error('Floodfill only implemented for rank-1, rank-2 and rank-3 tensors.');
+        }
+    }
+
+    /**
+     * Floodfill from boundary voxels below the given threshold.
+     * Only for rank-1, rank-2, and rank-3 tensors.
+     */
+    export function createFloodfilled(tensor: Tensor, threshold: number, mode: 'inside' | 'outside'): Tensor {
+        const { dataOffset: o } = tensor.space;
+        const visited = floodfill(tensor, threshold);
+        const fillValue = threshold * 2;
+
+        const floodfillGet = mode === 'outside'
+            ? (data: Tensor.Data, ...coords: number[]): number => {
+                const i = o(...coords);
```

Comment:
The `...` is a performance suicide here (it will do an array allocation for each call). This should have specialized per-dimension version.